File: app/src/main/java/com/usharik/app/DeclensionQuizState.java
package com.usharik.app;

import com.usharik.database.WordInfo;

/**
 * Holds the mutable state for a single full-declension quiz session.
 * Owned exclusively by DeclensionQuizViewModel.
 */
public class DeclensionQuizState {
    private WordInfo wordInfo;
    private WordTextModel[] wordTextModels;
    private String[][] correctAnswers;
    private int[][] actualAnswers;

    public DeclensionQuizState() {
        wordTextModels = new WordTextModel[14];
        correctAnswers = new String[2][7];
        actualAnswers = new int[2][7];
        for (int i = 0; i < 2; i++) {
            for (int j = 0; j < 7; j++) {
                actualAnswers[i][j] = -1;
            }
        }
    }

    public WordInfo getWordInfo() {
        return wordInfo;
    }

    public void setWordInfo(WordInfo wordInfo) {
        this.wordInfo = wordInfo;
    }

    public WordTextModel[] getWordTextModels() {
        return wordTextModels;
    }

    public void setWordTextModels(WordTextModel[] wordTextModels) {
        this.wordTextModels = wordTextModels;
    }

    public String[][] getCorrectAnswers() {
        return correctAnswers;
    }

    public int[][] getActualAnswers() {
        return actualAnswers;
    }

    public static class WordTextModel {
        private final String word;
        private int visible;

        public WordTextModel(String word, int visible) {
            this.word = word;
            this.visible = visible;
        }

        public String getWord() {
            return word;
        }

        public int getVisible() {
            return visible;
        }

        public void setVisible(int visible) {
            this.visible = visible;
        }
    }
}

File: app/src/main/kotlin/com/usharik/app/DeclensionQuizState.kt
package com.usharik.app

import com.usharik.database.WordInfo

/** Mutable session state, owned exclusively by DeclensionQuizViewModel. */
class DeclensionQuizState {
    private var wordInfo: WordInfo? = null
    private var wordTextModels = arrayOfNulls<WordTextModel>(14)
    private val correctAnswers = Array(2) { arrayOfNulls<String>(7) }
    private val actualAnswers = Array(2) { IntArray(7) { -1 } }
    private var wrongAttempts = 0

    fun getWordInfo() = wordInfo
    fun setWordInfo(value: WordInfo?) { wordInfo = value }
    fun getWordTextModels(): Array<WordTextModel?> = wordTextModels
    fun setWordTextModels(value: Array<WordTextModel?>) { wordTextModels = value }
    fun getCorrectAnswers(): Array<Array<String?>> = correctAnswers
    fun getActualAnswers(): Array<IntArray> = actualAnswers
    fun getWrongAttempts() = wrongAttempts
    fun incrementWrongAttempts() { wrongAttempts++ }
    fun resetWrongAttempts() { wrongAttempts = 0 }

    class WordTextModel(@JvmField val word: String, @JvmField var visible: Int) {
        fun getWord() = word
        fun getVisible() = visible
        fun setVisible(value: Int) { visible = value }
    }
}
